timemachine/fe/rbfe.py

```diff
@@ -459,6 +486,66 @@ def estimate_free_energy_given_initial_states(initial_states, protocol, temperat
     )
 
 
+def _optimize_coords_along_states(initial_states):
+    # use the end-state to define the optimization settings
+    end_state = initial_states[0]
+    ligand_coords = end_state.x0[end_state.ligand_idxs]
+    r_i = np.expand_dims(ligand_coords, axis=1)
+    r_j = np.expand_dims(end_state.x0, axis=0)
+    d_ij = np.linalg.norm(jax_utils.delta_r(r_i, r_j, box=end_state.box0), axis=-1)
+    cutoff = 0.5  # in nanometers
+    free_idxs = np.where(np.any(d_ij < cutoff, axis=0))[0].tolist()
+    x_opt = end_state.x0
+    x_traj = []
+    for initial_state in initial_states:
+        bound_impls = [p.bound_impl(np.float32) for p in initial_state.potentials]
+        val_and_grad_fn = minimizer.get_val_and_grad_fn(bound_impls, initial_state.box0, initial_state.lamb)
+        assert np.all(np.isfinite(x_opt)), "Initial coordinates contain nan or inf"
+        x_opt = minimizer.local_minimize(x_opt, val_and_grad_fn, free_idxs)
+        x_traj.append(x_opt)
+        assert np.all(np.isfinite(x_opt)), "Minimization resulted in a nan"
+        del bound_impls
+
+    return x_traj
+
+
+def optimize_coordinates(initial_states):
+    """
+    Optimize geometries of the initial states. The initial states are modified in place.
```

Comment:
Nit: Is it beneficial to modify `initial_states` in-place, rather than returning a copy?

(Don't see a performance advantage since copies are created within this function, do see examples of where it would be useful to compare coordinates before / after optimization, and in general find it easier to think about less state...)

Reply:
I'm going to go with a hybrid solution - just have this return a list of coordinates and have the calling function do the  modification (which does the setup) so it's more explicit.